src/java_bytecode/generic_recursion_map_keys.cpp

```diff
@@ -0,0 +1,158 @@
+/// Author: DiffBlue Limited. All Rights Reserved.
+
+/// \file
+/// Generic-recursion-map entries owner class. Generic-recursion-map maps
+/// generic parameters to a stack of their types in (every depth of) the
+/// current scope. This class adds entries to the generic-recursion-map
+/// for a particular scope, and ensures that they are erased on leaving that
+/// scope.
+
+#include "generic_recursion_map_keys.h"
+
+
+/// \param type Source type
+/// \return The vector of implicitly generic and (explicitly) generic type
+/// parameters of the given type.
+const std::vector<java_generic_parametert> get_all_generic_parameters
+  (const typet &type)
+{
+  PRECONDITION(is_java_generic_class_type(type) ||
+               is_java_implicitly_generic_class_type(type));
+  std::vector<java_generic_parametert> generic_parameters;
+  if(is_java_implicitly_generic_class_type(type))
+  {
+    const java_implicitly_generic_class_typet &implicitly_generic_class =
+      to_java_implicitly_generic_class_type(
+        to_java_class_type(type));
+    generic_parameters.insert(
+      generic_parameters.end(),
+      implicitly_generic_class.implicit_generic_types().begin(),
+      implicitly_generic_class.implicit_generic_types().end());
+  }
+  if(is_java_generic_class_type(type))
+  {
+    const java_generic_class_typet &generic_class =
+      to_java_generic_class_type(to_java_class_type(type));
+    generic_parameters.insert(
+      generic_parameters.end(),
+      generic_class.generic_types().begin(),
+      generic_class.generic_types().end());
+  }
+  return generic_parameters;
+}
+
+/// Creates a vector of pairs to be processed by function insert_pairs.
+/// \param parameters generic parameters that are the keys of the pairs to add
+/// \param types a type to add for each parameter
+/// \return a vector of pairs of a parameter and the corresponding type
+const std::vector<std::pair<java_generic_parametert, reference_typet>>
+pair_up(
+  const std::vector<java_generic_parametert> &parameters,
+  const std::vector<reference_typet> &types)
+{
+  PRECONDITION(!parameters.empty() && (parameters.size() == types.size()));
+  std::vector<std::pair<java_generic_parametert, reference_typet>> vector;
+  for(size_t i = 0; i < parameters.size(); i++)
+  {
+    vector.push_back(std::make_pair(parameters.at(i), types.at(i)));
+  }
+  return vector;
+}
+
+/// Add pairs to the controlled map. Own the keys and pop from their stack
+/// on destruction; otherwise do nothing.
+/// \param parameters generic parameters that are the keys of the pairs to add
+/// \param types a type to add for each parameter
+const void generic_recursion_map_keyst::insert_pairs(
+  const std::vector<java_generic_parametert> &parameters,
+  const std::vector<reference_typet> &types)
+{
+  INVARIANT(erase_keys.empty(), "insert_pairs should only be called once");
+  const std::vector<std::pair<java_generic_parametert, reference_typet>>
```

Comment:
I'd probably typedef `std::vector<std::pair<java_generic_parametert, reference_typet>>` somewhere

Reply:
I removed the function `pair_up` and the construction is done within `insert_pairs` so this type is explicitly used only once now.